File: RemoveInvalidParentheses301.java
class Solution {

    // TC:  O(2^n) - at every step, we are checking valid - not valid (Choose, don't choose) --> 2 power n
    // SC:  O(n)

    public List<String> removeInvalidParentheses(String s) {
        if(s == null)
            return new ArrayList<>();
        
        Queue<String> queue = new LinkedList<>();
        HashSet<String> set = new HashSet<>();
        List<String> output = new ArrayList<>();
        
        queue.add(s);
        set.add(s);
        
        String front = new String();
        String val = new String();
        boolean found = false;
        while(!queue.isEmpty()){
            front = queue.poll();
            
            if(isValid(front)){
                output.add(front);
                found = true;
            }
            
            if(!found){
                for(int x = 0; x < front.length(); x++){
                    // non braces
                    if(front.charAt(x) != ')' && front.charAt(x) != '('){
                        continue;
                    }

                    val = front.substring(0, x) + front.substring(x + 1, front.length());

                    if(!set.contains(val)){
                        set.add(val);
                        queue.add(val);
                    }
                
                }
            }
            
        }
        return output;
    }
    
    private boolean isValid(String str){
        int count = 0;
        for(int i = 0; i < str.length(); i++){
            if(str.charAt(i) != ')' && str.charAt(i) != '('){
                continue;
            }
            
            if(str.charAt(i) == '(')
                count += 1;
            
            if(count == 0 && str.charAt(i) == ')')
                return false;

            if(str.charAt(i) == ')')
                count -= 1;
        }
        return count == 0;
    }
}

File: removeInvalidParentheses_301.py
# Time Complexity : O(2^n), every point has choose/don't choose option
# Space Complexity : O(n)
# Did this code successfully run on Leetcode : Yes 
# Any problem you faced while coding this : No

# Approach/Intuition:

# We need all possible combinations: backtracking. We choose/don't choose a '(',')'.  
# Using DFS our code would explore each branch till we find an  answer and that would give us extra answers than required. But if we explore c/dc every bracket and validate the remaining string at every level, we won't have to explore more levels if we find our answer early on. We need to remove the minimum number of invalid parentheses in order to make the input string valid. 
# At the same level, when we find an answer, keep exploring the same level for more answers. If no answer is found, explore next level.
# Need to verify only unique strings: either keep a visited array or keep unique strings in a hashset
# Validating strings: Stack soln/We can count the number of ( and ) since we only have one type of brackets(+1-1+1... and so on). If count is negative at any point, return invalid string.
# When we encounter char, move the pointer ahead.

class Solution:
    def removeInvalidParentheses(self, s: str) -> List[str]:
        output = []
        
        if s==None:
            return output
        
        #1. put initial string in hashset and queue
        hset = set()
        queue = []
        found = False
        
        hset.add(s)
        queue.append(s)
        
        #2. start BFS
        while queue:
            front = queue.pop(0)
            #2.3  
            if (self.isValid(front)):
                found = True
                output.append(front)
            
            if not found:
                for x in range(len(front)):
                    if front[x] != ')' and front[x] != '(':
                        continue
                    newString = front[:x]+front[x+1:]
                    
                    #2.2
                    if newString not in hset:
                        hset.add(newString)
                        queue.append(newString)
            
            #2.1 Check if it is a () 
            #2.2 If it is present in hashset
            #2.3 check if valid
        
        return output
        
    def isValid(self, s):
        count = 0
        for ch in s:
            if ch == "(":
                count += 1
            elif ch == ")":
                if count == 0: return False
                count -= 1
        return count == 0
        
        